client/src/components/BulkCreateForm/BulkCreateForm.js

```diff
@@ -16,58 +16,154 @@ export default class BulkCreateForm extends React.Component {
+            // want to include it in the request
+            delete form.csv;
         }
+        return form;
+    }
+
+    submitForValidation() {
+        const { onValidationSubmit } = this.props;
+        onValidationSubmit(this.buildForm());
+    }
+
+    submitCreate() {
+        const { onCreateSubmit } = this.props;
+        onCreateSubmit(this.buildForm());
     }
 
     /**
      * Handles when a file is uploaded to the file upload input
      * @param event - The upload event fired by the input
      */
-    handleFileUploadChange(event) {
+    async handleFileUploadChange(event) {
         const { fileUpload } = this.state;
 
         this.resetForm(false);
 
         if (validator.matches(event.target.value, csvFileRegex)) {
-            fileUpload.file = event.target.files[0];
-            fileUpload.message = event.target.files[0].name;
-            fileUpload.fileType = '.csv';
+            fileUpload.csv = event.target.files[0];
             this.setState({fileUpload});
+            this.readCSVHeaders();
         }
         else if (validator.matches(event.target.value, zipFileRegex)) {
-            fileUpload.file = event.target.files[0];
-            fileUpload.message = event.target.files[0].name;
-            fileUpload.fileType = '.zip';
+            fileUpload.zip = event.target.files[0];
+            let content = await JSZip.loadAsync(fileUpload.zip);
+
+            for (let fileName in content.files) {
+                if (!content.files.hasOwnProperty(fileName) || fileName.startsWith('__MACOSX')) continue;
+                if (fileName.endsWith('.csv')) {
+                    if (fileUpload.csv) {
+                        fileUpload.errorMessage = 'Your .zip file contained multiple .csv files. Please only upload one .csv file at a time.';
+                        fileUpload.csv = null;
+                        fileUpload.zip = null;
+                        fileUpload.images = [];
+                        fileUpload.isValid = false;
+                        break;
+                    }
+                    fileUpload.csv = content.files[fileName];
+                } else {
+                    fileUpload.images.push(content.files[fileName]);
+                }
+            }
+
+            if (!fileUpload.csv && !fileUpload.errorMessage) {
+                fileUpload.errorMessage = 'Your zip file didn\'t contain a .csv file. Please check the contents of your .zip file and try again.';
+                fileUpload.zip = null;
+                fileUpload.images = [];
+                fileUpload.isValid = false;
+            }
+
             this.setState({fileUpload});
+            this.readCSVHeaders();
         }
         else {
             event.target.value = '';
             alert('Invalid file type submitted');
         }
     }
 
-    handleMoreInformationClick() {
-        const { showingMoreInformation } = this.state;
+    async readCSVHeaders() {
```

Comment:
Do you think we'll ever need to do any other CSV reading on the front-end and should extract some of this to a `util` function?

Reply:
I don't think we have any other use cases for it at the moment.